File: temporal_ts/src/webhook-bridge.ts
import http from 'node:http'
import { Client, Connection } from '@temporalio/client'

async function main() {
  const temporalAddress = process.env.TEMPORAL_ADDRESS || 'localhost:7233'
  const temporalNamespace = process.env.TEMPORAL_NAMESPACE || 'default'
  const port = Number(process.env.WEBHOOK_BRIDGE_PORT || '8088')

  const connection = await Connection.connect({ address: temporalAddress })
  const client = new Client({ connection, namespace: temporalNamespace })

  const server = http.createServer(async (req, res) => {
    if (req.url === '/health') {
      res.writeHead(200, { 'Content-Type': 'application/json' })
      res.end('{"ok":true}')
      return
    }

    if (req.url === '/webhooks' && req.method === 'POST') {
      try {
        const body = await readBody(req)
        const event = JSON.parse(body)
        const accountId = event.account || event?.data?.object?.account

        if (!accountId) {
          res.writeHead(200, { 'Content-Type': 'application/json' })
          res.end('{"status":"skipped","reason":"no account_id"}')
          return
        }

        const workflowIds = await findWorkflowsForAccount(client, accountId)
        let signaled = 0
        for (const wfId of workflowIds) {
          try {
            const handle = client.workflow.getHandle(wfId)
            await handle.signal('stripe_event', event)
            signaled++
          } catch {
            // Workflow not found — skip
          }
        }

        res.writeHead(200, { 'Content-Type': 'application/json' })
        res.end(JSON.stringify({ status: 'ok', signaled }))
      } catch (err: any) {
        if (err instanceof SyntaxError) {
          res.writeHead(400, { 'Content-Type': 'application/json' })
          res.end(JSON.stringify({ error: `Invalid JSON: ${err.message}` }))
        } else {
          res.writeHead(500, { 'Content-Type': 'application/json' })
          res.end(JSON.stringify({ error: err.message }))
        }
      }
      return
    }

    res.writeHead(404)
    res.end()
  })

  server.listen(port, () => {
    console.log('Starting webhook bridge...')
    console.log(`  Temporal: ${temporalAddress} (${temporalNamespace})`)
    console.log(`  Port:     ${port}`)
    console.log(`Webhook bridge listening on port ${port}`)
  })

  process.on('SIGINT', () => server.close())
  process.on('SIGTERM', () => server.close())
}

function readBody(req: http.IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    let body = ''
    req.on('data', (chunk) => {
      body += chunk
    })
    req.on('end', () => resolve(body))
    req.on('error', reject)
  })
}

async function findWorkflowsForAccount(client: Client, accountId: string): Promise<string[]> {
  try {
    const query = `WorkflowType = 'syncWorkflow' AND AccountId = '${accountId}' AND ExecutionStatus = 'Running'`
    const ids: string[] = []
    for await (const workflow of client.workflow.list({ query })) {
      ids.push(workflow.workflowId)
    }
    return ids
  } catch {
    return []
  }
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: infra/temporal_ruby/lib/webhook_bridge.rb
# frozen_string_literal: true

require 'temporalio/client'
require 'json'
require 'webrick'

# Webhook-to-signal bridge: receives Stripe webhooks and fires signals
# to matching SyncWorkflow instances.
#
# Routing: queries Temporal visibility for workflows with matching AccountId
# search attribute, then signals each one with the event payload.

class WebhookBridge
  def initialize(client:, port: 8088)
    @client = client
    @port = port
  end

  def start
    server = WEBrick::HTTPServer.new(Port: @port, Logger: WEBrick::Log.new($stdout, WEBrick::Log::INFO))

    server.mount_proc '/webhooks' do |req, res|
      handle_webhook(req, res)
    end

    server.mount_proc '/health' do |_req, res|
      res.status = 200
      res.body = '{"ok":true}'
      res['Content-Type'] = 'application/json'
    end

    trap('INT') { server.shutdown }
    trap('TERM') { server.shutdown }

    puts "Webhook bridge listening on port #{@port}"
    server.start
  end

  private

  def handle_webhook(req, res)
    event = JSON.parse(req.body)

    account_id = event['account'] || event.dig('data', 'object', 'account')

    unless account_id
      res.status = 200
      res.body = '{"status":"skipped","reason":"no account_id"}'
      res['Content-Type'] = 'application/json'
      return
    end

    workflow_ids = find_workflows_for_account(account_id)

    signaled = 0
    workflow_ids.each do |wf_id|
      handle = @client.workflow_handle(wf_id)
      handle.signal('stripe_event', event)
      signaled += 1
    rescue Temporalio::Error::WorkflowNotFoundError
      next
    end

    res.status = 200
    res.body = JSON.generate(status: 'ok', signaled: signaled)
    res['Content-Type'] = 'application/json'
  rescue JSON::ParserError => e
    res.status = 400
    res.body = JSON.generate(error: "Invalid JSON: #{e.message}")
    res['Content-Type'] = 'application/json'
  rescue StandardError => e
    res.status = 500
    res.body = JSON.generate(error: e.message)
    res['Content-Type'] = 'application/json'
  end

  def find_workflows_for_account(account_id)
    query = "WorkflowType = 'SyncWorkflow' AND AccountId = '#{account_id}' AND ExecutionStatus = 'Running'"
    workflows = @client.list_workflows(query: query)
    workflows.map(&:id)
  rescue StandardError
    []
  end
end

if __FILE__ == $PROGRAM_NAME
  temporal_address = ENV.fetch('TEMPORAL_ADDRESS', 'localhost:7233')
  temporal_namespace = ENV.fetch('TEMPORAL_NAMESPACE', 'default')
  port = ENV.fetch('WEBHOOK_BRIDGE_PORT', '8088').to_i

  client = Temporalio::Client.connect(temporal_address, temporal_namespace)

  puts "Starting webhook bridge..."
  puts "  Temporal: #{temporal_address} (#{temporal_namespace})"
  puts "  Port:     #{port}"

  WebhookBridge.new(client: client, port: port).start
end
